packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/utils.ts

```diff
@@ -185,3 +139,117 @@ export const allocateQuantities =
+
+    toAllocate -= allocatedNumberOfPacks * draftOutboundLine.packSize;
+
+    newDraftOutboundLines[draftOutboundLineIdx] = {
+      ...draftOutboundLine,
+      numberOfPacks: allocatedNumberOfPacks,
+    };
+  });
+  return toAllocate;
+};
+
+const allocatedAdditionalStock = ({
+  validBatches,
+  newDraftOutboundLines,
+  toAllocate,
+}: {
+  validBatches: DraftOutboundLine[];
+  newDraftOutboundLines: DraftOutboundLine[];
+  toAllocate: number;
+}) => {
+  validBatches.forEach(batch => {
+    const draftOutboundLineIdx = newDraftOutboundLines.findIndex(
+      ({ id }) => batch.id === id
+    );
+    const draftOutboundLine = newDraftOutboundLines[draftOutboundLineIdx];
+    if (!draftOutboundLine) return null;
+    if (toAllocate < 0) return null;
+    if (draftOutboundLine.packSize > toAllocate) return null;
+
+    const availableUnits =
+      ((draftOutboundLine.stockLine?.availableNumberOfPacks ?? 0) -
+        draftOutboundLine.numberOfPacks) *
+      draftOutboundLine.packSize;
+
+    if (availableUnits <= 0) return null;
+
+    const unitsToAllocate = Math.min(toAllocate, availableUnits);
+    const numberOfPacks = unitsToAllocate / draftOutboundLine.packSize;
+    const allocatedNumberOfPacks = Math.floor(numberOfPacks);
+
+    toAllocate -= allocatedNumberOfPacks * draftOutboundLine.packSize;
+    draftOutboundLine.numberOfPacks = allocatedNumberOfPacks;
+  });
+  return toAllocate;
+};
+
+const reduceBatchAllocation = ({
+  toAllocate,
+  validBatches,
+  newDraftOutboundLines,
+  includeOversizePacks = false,
+}: {
+  toAllocate: number;
+  validBatches: DraftOutboundLine[];
+  newDraftOutboundLines: DraftOutboundLine[];
+  includeOversizePacks?: boolean;
+}) => {
+  validBatches
+    .slice()
+    .sort(SortUtils.byExpiryDesc)
+    .forEach(batch => {
+      const draftOutboundLineIdx = newDraftOutboundLines.findIndex(
+        ({ id }) => batch.id === id
+      );
+      const draftOutboundLine = newDraftOutboundLines[draftOutboundLineIdx];
+      if (!draftOutboundLine) return null;
+      if (includeOversizePacks && draftOutboundLine.packSize < toAllocate)
+        return null;
+      if (!includeOversizePacks && draftOutboundLine.packSize > toAllocate)
+        return null;
+      if (draftOutboundLine.numberOfPacks === 0) return null;
+
+      const allocatedUnits =
+        draftOutboundLine.numberOfPacks * draftOutboundLine.packSize;
+      const unitsToReduce = Math.min(toAllocate, allocatedUnits);
+      const numberOfPacks = Math.floor(
+        (allocatedUnits - unitsToReduce) / draftOutboundLine.packSize
+      );
+      toAllocate -= includeOversizePacks
+        ? (draftOutboundLine.numberOfPacks - numberOfPacks) *
+          draftOutboundLine.packSize
+        : unitsToReduce;
+
+      newDraftOutboundLines[draftOutboundLineIdx] = {
+        ...draftOutboundLine,
+        numberOfPacks: numberOfPacks,
+      };
+    });
+  return Math.abs(toAllocate);
+};
```

Comment:
This is all very awesome and a great clean up of some very annoying logic - nice